### ISSUE TITLE
Packaging, Environments

### ISSUE THREAD

**Issue by kousu** [shimming-toolbox/shimming-toolbox#369]
### Description of Problem / Inefficiency:

There are several points that could be improved with the installation process.

#### Short term action tasks:

- [ ] avoid long conda solver times: (PR: https://github.com/shimming-toolbox/fsleyes-plugin-shimming-toolbox/pull/36)
  - maybe figure out if there's a dependency that's causing the problem
  - try to find a conda config file we could tweak to make it faster
- [x] don't downgrade wxPython (PR: https://github.com/shimming-toolbox/fsleyes-plugin-shimming-toolbox/issues/34)
  - solve the bugs with wxPython==4.1.1 
  - The reason for the downgrade has been lost when moving the plugin from [ST repository](https://github.com/shimming-toolbox/shimming-toolbox) to the [plugin repository](https://github.com/shimming-toolbox/fsleyes-plugin-shimming-toolbox)

- [x] drop st_venv and just use the base conda, to save download time and make it the deployed environment cleaner (PRs: https://github.com/shimming-toolbox/fsleyes-plugin-shimming-toolbox/pull/37/, https://github.com/shimming-toolbox/shimming-toolbox/pull/379)
  - problem: need to download an installer with the version of python you want to use
    - https://docs.conda.io/en/latest/miniconda.html#linux-installers
    - https://docs.conda.io/en/latest/miniconda.html#macos-installers
- [x] ditto for pst_venv -> base conda (PRs: ditto, we just did both in one day)
  - needs solving any dependency conflicts
  - there may be very very many

#### Medium term action tasks:

- look into psutil compilation errors at pip install
- drop the Makefile + installer scripts; replace with a pip package(s) -> publish to pypi
  - can't explicitly depend on dcm2niix
  
- don't assume you've installed conda
   - we can keep a helpful user 101 installer script that does `curl conda-installer.sh && sh conda-installer.sh && activate conda && pip install shimming-toolbox && create_launchers()`, but for people with pre-existing conda installs they can choose to avoid the redundancy/conflict with our conda and just use theirs (or just use `pip install --user`)

#### Long term action tasks:

- Write a conda package -> publish to conda-forge
  - can explicitly depend on dcm2niix, and maybe prelude
  -

**Comment by kousu** [comment]
Here's a diagram we drew during the meeting today, mapping out the dependencies we have:

![dependencies](https://user-images.githubusercontent.com/987487/159806324-2010e750-7fe4-4ba5-9935-f40bb2c0d345.png)

It's uh, a bit incomplete because we were talking it over as I was drawing it. To fill in the gaps:

* currently we ship as [`make install`](https://github.com/shimming-toolbox/shimming-toolbox/blob/6ed65f41b11ad9c727b02d098c82a6dab43a6355/Makefile#L12-L24) (+ [several bash scripts](https://github.com/shimming-toolbox/shimming-toolbox/tree/6ed65f4/installer))
* we use [conda to install `dcm2niix`](https://github.com/shimming-toolbox/shimming-toolbox/blob/6ed65f41b11ad9c727b02d098c82a6dab43a6355/installer/install_shimming_toolbox.sh#L46), the main non-python dependency
* we have documentation instructing the user [to install `prelude`](https://github.com/shimming-toolbox/shimming-toolbox/blob/4a99cec342ea4e52b164ad97ad96c93acad2ea64/docs/source/user_section/installation.rst#dependencies), the other non-python dependency; but we consider that optional
* we use [pip to install](https://github.com/shimming-toolbox/shimming-toolbox/blob/6ed65f41b11ad9c727b02d098c82a6dab43a6355/installer/install_shimming_toolbox.sh#L51) [all other dependencies](https://github.com/shimming-toolbox/shimming-toolbox/blob/6ed65f41b11ad9c727b02d098c82a6dab43a6355/setup.py#L36-L53) (meaning all are python code)
* The [GUI](https://github.com/shimming-toolbox/fsleyes-plugin-shimming-toolbox/) has the same structure

    With the addition that it ["declares" its dependency on the CLI with `curl` + `make`](https://github.com/shimming-toolbox/fsleyes-plugin-shimming-toolbox/blob/b2f2f598a4a520bc60e6e5de3e2de04e3ca05132/installer/install_shimming_toolbox.sh#L17-L30)

* Over in the [GUI](https://github.com/shimming-toolbox/fsleyes-plugin-shimming-toolbox/), we [downgrade wxPython](https://github.com/shimming-toolbox/fsleyes-plugin-shimming-toolbox/blob/b2f2f598a4a520bc60e6e5de3e2de04e3ca05132/installer/install_plugin.sh#L44-L46) from its current latest 4.1.1 to 4.0, to fix unspecified "bugs" which complicates things (and slows down the install)
    * Also, wxPython [doesn't have linux binaries](https://pypi.org/project/wxPython/#files) -- only macOS and Windows -- because it's tricky. [fslpy ran into the same problem](https://git.fmrib.ox.ac.uk/fsl/fslpy/-/blob/afb2f31d/README.rst#fslpy) and recommended `pip install -f https://extras.wxpython.org/wxPython4/extras/linux/gtk2/ubuntu-16.04/ wxpython` -- but that has a lot of pitfalls like: loading the wrong `libgtk.so`, needing the user to know what distro they're on, etc etc. `conda` on the other hand [does package wxPython binaries for all platforms](https://anaconda.org/conda-forge/wxpython/files), which encourages us towards getting it from conda

* Oh and [`dcm2bids`](https://github.com/UNFmontreal/Dcm2Bids) has an undeclared dependency on `dcm2niix` as well -- for the same reason: it's not python code so it can't easily become a python dependency. [They recommend creating a conda environment to get it](https://unfmontreal.github.io/Dcm2Bids/docs/2-tutorial/#setup).

**Comment by kousu** [comment]
We spent an hour trying to speed up conda's install. We learned:

- the issue is triggered for us only on Linux when we downgrade wxpython `conda install -c conda-forge wxpython=4.0.7`; there must be some incompatible setup (e.g. installing python=3.7 currently installs python=3.7.13=h12debd9_0, but the available wxpython builds [on conda-forge](https://anaconda.org/conda-forge/wxpython/files?) on linux seem to need either "python=3.7.8=h8bdb77d_2_cpython" or "python=3.7.11=h88f2d9e_0 ".

    We weren't able to reliably identify which packages cause the incompatibilities, despite adding many `-v` flags.
- the current version mixture on macOS installs quickly, so there must just be a compatible mixture that happens to exist on conda-forge at the moment
- if you run `conda instal -vvv ....` you will see some extra details, especially lines about the SAT solver it insists on invoking:

    ```
    DEBUG conda.resolve:generate_spec_constraints(924): generate_spec_constraints returning with clause count: 463567
    ```

    which at least lets you have an idea of how complicated the requested combination of packages is
- There are people complaining about this issue:
  - https://github.com/conda/conda/issues/10383 - no answer
  - https://github.com/conda/conda/issues/8051 - many answers, but none seemed to help us:
    - switch to [mamba](https://github.com/mamba-org/mamba)
    - run `conda config --remove channels conda-forge; conda config --add channels conda-forge; conda config --remove channels conda-forge; conda config --add channels conda-forge;`
    - add more RAM(?!)
    - reinstall conda
    - `rm ~/.condarc`
    - `conda config --set channel_priority flexible` (contradicting conda's own advice)
   
  None of these solutions worked for us.
- https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-channels.html recommends

    ```
    conda config --set channel_priority strict   
    ``` 

    >  Strict channel priority can dramatically speed up conda operations and also reduce package incompatibility problems. We recommend it as a default

    However we tried it and it actually made the macOS install *slower*.


We are instead going to focus on fixing the wx bugs so that we don't need to downgrade it at all, and hope incompatibilities and long solver times never come up again, because if they do we will not know how to solve them.

**Comment by po09i** [comment]
I opened a PR to fix the wx bugs here: https://github.com/shimming-toolbox/fsleyes-plugin-shimming-toolbox/pull/35

It seems rather straightforward to get working.

**Comment by kousu** [comment]
> I opened a PR to fix the wx bugs here: [shimming-toolbox/fsleyes-plugin-shimming-toolbox#35](https://github.com/shimming-toolbox/fsleyes-plugin-shimming-toolbox/pull/35)
> 
> It seems rather straightforward to get working.

Nice!

In the meantime, I had an insight and realized I hadn't used mamba correctly yesterday; using it correctly **does** solve the long solver time: https://github.com/shimming-toolbox/fsleyes-plugin-shimming-toolbox/pull/36.

If you've already solved the wx bugs then maybe you don't even need that at all. My long term suggestion is still to write basic no-frills conda and pip *packages* and then generally not force a choice of environment manager onto the users, so ideally there wouldn't be any mention of 'conda' or 'mamba' (or docker or podman or vagrant or virtualenv or pipenv or anything) in this codebase.

But in the meantime, using `conda` is always going to risk someone getting bit by its SAT solver when trying to deploy it on a platform we haven't tested -- either newer or older, anything sitting slightly outside of the mainstream collection of packages on conda-forge.

**Comment by kousu** [comment]
@gaspardcereza noted it's pretty hard to explain to users why the plugin depends on shimming-toolbox and not the other way around, when it seems like a sub-feature of the project.

The reason it's like this is because fsleyes is a pain. Installing fsleyes [is possible via pip](https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/FSLeyes#Install_from_PyPi_.28advanced.29) but is [better via conda](https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/FSLeyes#Install_from_conda-forge_.28recommended.29) due to very painful wxPython compatibility glitches (like ["ImportError: libgtk-x11-2.0.so.0: cannot open shared object file: No such file or directory"](https://github.com/shimming-toolbox/fsleyes-plugin-shimming-toolbox/issues/34#issuecomment-1085065013)).

We also ran into glitches where conda got confused about what conda to use, if a user has a pre-existing conda install.

But it would be ideal if the plugin was instead an _optional dependency_ of the main project.

## pip optional dependencies

We just confirmed that `pip` recognizes `fsleyes` as a package, even if installed via `conda`, and vice versa. This was not the case a year ago when the project started and fsleyes wasn't packaged for conda or pip. That means we could write fsleyes as an optional **pip** dependency:

```
# setup.py
# 
    extras_require={
        "GUI": ["fsleyes-plugin-shimming-toolbox"],
    },
#....
```

and

```
# setup.py  in fsleyes-plugin-shimming-toolbox
# 

    install_requires=[
     "fsleyes~=1.3"],
    ],
#....
```

This would invert the dependency diagram I drew last week.


Then:

For users unfamiliar with the command line or conda and who've never used fsleyes, shorten the installer script to:

```
# install_shimming_toolbox.sh

curl conda 
install_conda.sh ...

$ST_DIR/python/bin/conda install fsleyes=1.3.3

$ST_DIR/python/bin/pip install -y shimming_toolbox[gui]  #<-- will recognize that fsleyes is installed
```

While people are familiar with the command line and want to do large batch processing and who don't care about the GUI:

```
pip install shimming-toolbox
```

People who already have fsleyes installed, with or without conda:

```
pip install shimming-toolbox[gui]  # and pray
```

People who like conda and already use it:

```
conda env create shimming && conda activate shimming # optional 
conda install fsleyes
pip install shimming-toolbox[gui]
```


People who care about the GUI but know enough about the command line to not want to use conda, first direct users to https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/FSLeyes#Install_from_PyPi_.28advanced.29 and then they will run:

```
pip install -f https://extras.wxpython.org/wxPython4/extras/linux/gtk2/$DISTRO wxpython
pip install shimming-toolbox[gui]  # and pray a bit
```

**tl;dr**: we tell users they have two options:

1. `pip install shimming-toolbox`
2. `pip install shimming-toolbox[gui]`

and either that just works for them or they can probably figure out the rest.

## conda-based optional dependencies

It's unclear to me how users use this. There's no `conda install --optional`? https://github.com/conda/conda/issues/1696#issuecomment-300309736 says you add this to your..conda environment.yml??

```
requirements:
  run_constrained:
    - package 1.0.*
    - another-package >=2
 ```

I don't really get it. Docs didn't go in with the original feature :/ : https://github.com/conda/conda-build/issues/1964


Maybe it's just that users run:

```
conda install shimming-toolbox fsleyes-plugin-shimming-toolbox
```

and then what that means is...`conda remove fsleyes-plugin-shimming-toolbox` also removes `shimming-toolbox`, but not much else.

**Comment by kousu** [comment]
I found another cause for the long installer times: mixing conda-forge and anaconda. If you just install conda from https://github.com/conda-forge/miniforge/releases, which has anaconda's channels entirely removed, the install never hits "Retrying with flexible solve."

**Comment by kousu** [comment]
And it's the same story for `fsleyes`, which is installed by https://github.com/shimming-toolbox/fsleyes-plugin-shimming-toolbox/'s install script.

**Comment by kousu** [comment]
On long conda solver times: we spent an entire session trying different options to try to isolate and went in circles. 

After dropping wxPython==3.0.7 (https://github.com/shimming-toolbox/fsleyes-plugin-shimming-toolbox/pull/35) and the extraneous conda envs (#379), I'm still seeing a 5min+ install time on Linux -- both my personal Arch laptop and the older Ubuntu 20.04 LTS install on our internal servers -- while @po09i sees a much better 40s install time on macOS, and most of that is because the linux installs get stuck at conda's solver, but also now we hardly `conda install` anything anymore, just dcm2niix and the python version to install. Which means maybe we will have better luck exploring the space of possible factors. We could try:

a. removing the version constraint on dcm2niix
b. bumping python to 3.8
c. bumping python to 3.9
d. running `conda update -n base -c defaults conda` to make sure everything is latest and therefore most likely to be well maintained? idk

https://github.com/shimming-toolbox/fsleyes-plugin-shimming-toolbox/pull/36 sidesteps the issue (it's unfinished but wouldn't take long to finish properly) by using a faster solver algorithm, so it's one solution, but maybe we'll have better luck isolating the cause of the solver problem.

I think we can fit in these experiments for next week?

**Comment by kousu** [comment]
Ah a silly bug, now after #379 if I have a left-over install it doesn't reset the python version:

<details>

```
[kousu@nigiri fsleyes-plugin-shimming-toolbox]$ make install
/home/kousu/shimming-toolbox and conda install found, skipping install of conda

Installing fsleyes
[...]
## Package Plan ##

  environment location: /home/kousu/shimming-toolbox/python

  added / updated specs:
    - fsleyes=1.3.3

[...]
The following packages will be UPDATED:

  conda              pkgs/main::conda-4.11.0-py39h06a4308_0 --> conda-forge::conda-4.12.0-py39hf3d152e_0
  openssl                                 1.1.1m-h7f8727e_0 --> 1.1.1n-h7f8727e_0

The following packages will be SUPERSEDED by a higher-priority channel:

  ca-certificates    pkgs/main::ca-certificates-2021.10.26~ --> conda-forge::ca-certificates-2021.10.8-ha878542_0
  certifi            pkgs/main::certifi-2021.10.8-py39h06a~ --> conda-forge::certifi-2021.10.8-py39hf3d152e_2

[...]
Requirement already satisfied: numpy in /home/kousu/shimming-toolbox/python/lib/python3.9/site-packages (from imageio->fsleyes-plugin-shimming-toolbox==0.0.0) (1.21.1)
[...]
```
</details>

Notice that it installed to `python3.9/site-packages`, not `python3.7/site-packages`.

<details>

```
[kousu@nigiri fsleyes-plugin-shimming-toolbox]$ shimming-toolbox
[kousu@nigiri fsleyes-plugin-shimming-toolbox]$ Traceback (most recent call last):
  File "/home/kousu/shimming-toolbox/python/bin/fsleyes", line 7, in <module>
    from fsleyes.filtermain import main
  File "/home/kousu/shimming-toolbox/python/lib/python3.9/site-packages/fsleyes/__init__.py", line 367, in <module>
    from fsleyes.main import embed, shutdown  # noqa
  File "/home/kousu/shimming-toolbox/python/lib/python3.9/site-packages/fsleyes/main.py", line 50, in <module>
    import fsleyes.cliserver  as cliserver
  File "/home/kousu/shimming-toolbox/python/lib/python3.9/site-packages/fsleyes/cliserver.py", line 37, in <module>
    import fsleyes.actions.applycommandline as applycli
  File "/home/kousu/shimming-toolbox/python/lib/python3.9/site-packages/fsleyes/actions/__init__.py", line 106, in <module>
    import fsleyes_props   as props
  File "/home/kousu/shimming-toolbox/python/lib/python3.9/site-packages/fsleyes_props/__init__.py", line 319, in <module>
    from .properties_types import (
  File "/home/kousu/shimming-toolbox/python/lib/python3.9/site-packages/fsleyes_props/properties_types.py", line 36, in <module>
    import matplotlib.pyplot as plt
  File "/home/kousu/shimming-toolbox/python/lib/python3.9/site-packages/matplotlib/__init__.py", line 205, in <module>
    _check_versions()
  File "/home/kousu/shimming-toolbox/python/lib/python3.9/site-packages/matplotlib/__init__.py", line 190, in _check_versions
    from . import ft2font
ImportError: /home/kousu/shimming-toolbox/python/lib/python3.9/site-packages/matplotlib/ft2font.cpython-39-x86_64-linux-gnu.so: undefined symbol: _ZSt28__throw_bad_array_new_lengthv
```

</details>
